sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -146,54 +145,84 @@ void Scheduler::GraphBuilder::UpdateLeafs(
   }
 }
 
-void Scheduler::GraphBuilder::AddNodeToLeafs(
-    MemObjRecord *Record, Command *Cmd, Requirement *Req) {
-  if (Req->MAccessMode == access::mode::read)
+void Scheduler::GraphBuilder::AddNodeToLeafs(MemObjRecord *Record, Command *Cmd,
+                                             access::mode AccessMode) {
+  if (AccessMode == access::mode::read)
     Record->MReadLeafs.push_back(Cmd);
   else
     Record->MWriteLeafs.push_back(Cmd);
 }
 
+UpdateHostRequirementCommand *Scheduler::GraphBuilder::insertUpdateHostReqCmd(
+    MemObjRecord *Record, Requirement *Req, const QueueImplPtr &Queue) {
+  AllocaCommandBase *AllocaCmd = findAllocaForReq(Record, Req, Queue);
+  assert(AllocaCmd && "There must be alloca for requirement!");
+  UpdateHostRequirementCommand *UpdateCommand =
+      new UpdateHostRequirementCommand(Queue, Req, AllocaCmd);
+  // Need copy of requirement because after host accessor destructor call
+  // dependencies become invalid if requirement is stored by pointer.
+  Requirement *StoredReq = UpdateCommand->getStoreRequirement();
+
+  std::set<Command *> Deps = findDepsForReq(Record, Req, Queue);
+  for (Command *Dep : Deps) {
+    UpdateCommand->addDep({Dep, StoredReq, AllocaCmd});
+    Dep->addUser(UpdateCommand);
+  }
+  // access::mode::read_write is always used here regardless requieremnt
+  // access mode because this node shouldn't be skipped.
+  UpdateLeafs(Deps, Record, access::mode::read_write);
+  AddNodeToLeafs(Record, UpdateCommand, access::mode::read_write);
+  return UpdateCommand;
+}
+
 MemCpyCommand *
 Scheduler::GraphBuilder::insertMemCpyCmd(MemObjRecord *Record, Requirement *Req,
                                          const QueueImplPtr &Queue,
                                          bool UseExclusiveQueue) {
-
-  Requirement FullReq(/*Offset*/ {0, 0, 0}, Req->MMemoryRange,
-                      Req->MMemoryRange, access::mode::read_write,
-                      Req->MSYCLMemObj, Req->MDims, Req->MElemSize);
-
-  std::set<Command *> Deps = findDepsForReq(Record, &FullReq, Queue);
+  std::set<Command *> Deps = findDepsForReq(Record, Req, Queue);
   QueueImplPtr SrcQueue = (*Deps.begin())->getQueue();
-  AllocaCommandBase *AllocaCmdDst = findAllocaForReq(Record, &FullReq, Queue);
 
-  if (!AllocaCmdDst) {
-    std::unique_ptr<AllocaCommand> AllocaCmdUniquePtr(
-        new AllocaCommand(Queue, FullReq));
-
-    if (!AllocaCmdUniquePtr)
-      throw runtime_error("Out of host memory");
-
-    Record->MAllocaCommands.push_back(AllocaCmdUniquePtr.get());
-    AllocaCmdDst = AllocaCmdUniquePtr.release();
-    Deps.insert(AllocaCmdDst);
+  AllocaCommandBase *AllocaCmdDst = getOrCreateAllocaForReq(Record, Req, Queue);
+  if (!AllocaCmdDst)
+    throw runtime_error("Out of host memory");
+  Deps.insert(AllocaCmdDst);
+  // Get parrent alloca of sub buffer to perform full copy of whole buffer
```

Comment:
`parent` and `sub-buffer`.

Otherwise I have a problem every time I look at `alloca` in this code because I am confused by http://man7.org/linux/man-pages/man3/alloca.3.html or https://llvm.org/docs/LangRef.html#alloca-instruction

What about using something different? :-) such as use the real name, "allocation" in comments, and shorter abbreviations in the code, such as `AllocCmdDst`...

Reply:
Made the comment more clear according to your suggestion.